cities/queries/causal_insight.py

```diff
@@ -0,0 +1,413 @@
+import os
+
+import dill
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import plotly.graph_objects as go
+import pyro
+import torch
+
+from cities.modeling.model_interactions import model_cities_interaction
+from cities.modeling.modeling_utils import (
+    prep_wide_data_for_inference,
+    revert_prediction_df,
+    revert_standardize_and_scale_scaler,
+)
+from cities.utils.cleaning_utils import find_repo_root
+from cities.utils.data_grabber import DataGrabber
+
+
+class CausalInsight:
+    def __init__(self, outcome_dataset, intervention_dataset, num_samples=1000):
+        self.outcome_dataset = outcome_dataset
+        self.intervention_dataset = intervention_dataset
+        self.root = find_repo_root()
+        self.num_samples = num_samples
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def load_guide(self, forward_shift):
+        pyro.clear_param_store()
+        guide_name = (
+            f"{self.intervention_dataset}_{self.outcome_dataset}_{forward_shift}"
+        )
+        guide_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_guide.pkl"
+        )
+
+        with open(guide_path, "rb") as file:
+            self.guide = dill.load(file)
+        param_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_params.pth"
+        )
+
+        pyro.get_param_store().load(param_path)
+
+        self.forward_shift = forward_shift
+
+    def generate_samples(self):
+        self.data = prep_wide_data_for_inference(
+            outcome_dataset=self.outcome_dataset,
+            intervention_dataset=self.intervention_dataset,
+            forward_shift=self.forward_shift,
+        )
+        self.model = model_cities_interaction
+
+        self.model_args = self.data["model_args"]
+
+        self.model_conditioned = pyro.condition(
+            self.model,
+            data={"T": self.data["t"], "Y": self.data["y"], "X": self.data["x"]},
+        )
```

Comment:
Is this necessary here? Cz there's no inference happening.

Reply:
good catch, this was a redundant piece that was used in an earlier version. Removed.